File: unplot/curves/vectorpaths.py
"""Pure curve geometry: turn (possibly tangled) draw-order polylines into clean single-valued curves.

No fitz, no domain knowledge: numpy in, numpy out. These are the de-fan / chain / lobe-split operations
that recover individual curves when a PDF draws several lobes as one continuous pen stroke with jumps and
returns between them.

Orientation convention: source y increases DOWNWARD (PDF/image), so visual 'up' = decreasing y. `split_lobes`
treats `-y` as height accordingly.
"""
from __future__ import annotations

import numpy as np
from numpy.typing import NDArray


def x_sorted(P: NDArray) -> NDArray:
    return P[np.argsort(P[:, 0])]


def mono_x_segments(P: NDArray) -> list[NDArray]:
    """Split a draw-order polyline into maximal monotone-in-x runs; orient each x-increasing."""
    segs: list[NDArray] = []
    n = len(P)
    s = 0
    dirn = 0
    for i in range(1, n):
        d = P[i, 0] - P[i - 1, 0]
        if abs(d) < 1e-9:
            continue
        sd = 1 if d > 0 else -1
        if dirn == 0:
            dirn = sd
        elif sd != dirn:
            segs.append(P[s:i])
            s = i - 1
            dirn = sd
    segs.append(P[s:n])
    return [seg[::-1] if seg[-1, 0] < seg[0, 0] else seg for seg in segs]


def _is_valley_join(cur: NDArray, t: NDArray, k: int = 8, margin: float = 2.0) -> bool:
    """True if joining cur's END to t's START sits at a VALLEY — the join point is lower in data (higher
    source-y) than its neighbours a few points back in `cur` and forward in `t`. That's two lobes meeting at
    the baseline, not a broken stroke continuing the same way (where the neighbours straddle the join). Refusing
    valley joins stops adjacent lobes merging into one full-width curve, while still rejoining broken strokes."""
    n = len(cur)
    yb = float(np.mean(cur[max(0, n - k - 1):n - 1, 1])) if n > 1 else float(cur[-1, 1])
    ya = float(np.mean(t[1:k + 1, 1])) if len(t) > 1 else float(t[0, 1])
    ey = float(cur[-1, 1])
    return ey > yb + margin and ey > ya + margin


def chain_curves(segs: list[NDArray], xtol: float = 4.0, ytol: float = 6.0) -> list[NDArray]:
    """Chain x-increasing segments where one's END ~ next's START (both x and y). Rebuilds a curve whose
    stroke the PDF broke at a crossing (its descending tail orphaned into the next path object) — but refuses
    a join that sits at a valley, so two adjacent lobes meeting at the baseline are not merged into one."""
    segs = sorted(segs, key=lambda s: s[0, 0])
    used = [False] * len(segs)
    curves: list[NDArray] = []
    for i, s in enumerate(segs):
        if used[i]:
            continue
        cur = s.copy()
        used[i] = True
        ex, ey = cur[-1]
        ch = True
        while ch:
            ch = False
            for j, t in enumerate(segs):
                if used[j]:
                    continue
                if abs(t[0, 0] - ex) < xtol and abs(t[0, 1] - ey) < ytol and not _is_valley_join(cur, t):
                    cur = np.vstack([cur, t])
                    used[j] = True
                    ex, ey = cur[-1]
                    ch = True
                    break
        curves.append(cur)
    return curves


def split_lobes(curve: NDArray, depth_frac: float = 0.35, min_pts: int = 12) -> list[NDArray]:
    """Split a single-valued curve into separate lobes at DEEP inter-lobe valleys (near baseline), while
    keeping shallow within-lobe dips (real double-humps) intact. `h = -y` (visual height)."""
    h = -curve[:, 1]
    h = h - h.min()
    if h.max() <= 0:
        return [curve]
    peaks = [i for i in range(1, len(h) - 1)
             if h[i] >= h[i - 1] and h[i] >= h[i + 1] and h[i] > 0.15 * h.max()]
    if len(peaks) < 2:
        return [curve]
    splits: list[int] = []
    for a, b in zip(peaks, peaks[1:]):
        v = a + int(np.argmin(h[a:b + 1]))
        if h[v] < depth_frac * min(h[a], h[b]):
            splits.append(v)
    if not splits:
        return [curve]
    out: list[NDArray] = []
    prev = 0
    for v in splits:
        if v - prev >= min_pts:
            out.append(curve[prev:v + 1])
        prev = v
    if len(curve) - prev >= min_pts:
        out.append(curve[prev:])
    return out

File: js/src/curves/vectorpaths.ts
/**
 * Pure curve geometry — TS port of the Python `curves/vectorpaths.py`. No dependencies, plain array math:
 * de-fan a draw-order polyline into monotone-x runs, chain runs back into curves (refusing valley joins so
 * adjacent lobes don't merge), and split a curve at deep inter-lobe valleys.
 *
 * Convention: points are [x, y] in source coords with y increasing DOWNWARD (so "visual up" = decreasing y),
 * matching the pdf.js adapter output.
 */
export type Pt = [number, number];

const mean = (a: number[]): number => a.reduce((s, x) => s + x, 0) / a.length;

export const xSorted = (P: Pt[]): Pt[] => [...P].sort((a, b) => a[0] - b[0]);

/** Split a draw-order polyline into maximal monotone-in-x runs; orient each x-increasing. */
export function monoXSegments(P: Pt[]): Pt[][] {
  const segs: Pt[][] = [];
  let s = 0, dir = 0;
  for (let i = 1; i < P.length; i++) {
    const d = P[i]![0] - P[i - 1]![0];
    if (Math.abs(d) < 1e-9) continue;
    const sd = d > 0 ? 1 : -1;
    if (dir === 0) dir = sd;
    else if (sd !== dir) { segs.push(P.slice(s, i)); s = i - 1; dir = sd; }
  }
  segs.push(P.slice(s));
  return segs.map((seg) => (seg[seg.length - 1]![0] < seg[0]![0] ? [...seg].reverse() : seg));
}

/** True if joining cur's END to t's START sits at a VALLEY (the join is lower in data than its neighbours both
 *  back and forward) — two lobes meeting at the baseline, not a broken stroke continuing the same way. */
function isValleyJoin(cur: Pt[], t: Pt[], k = 8, margin = 2.0): boolean {
  const n = cur.length;
  const yb = n > 1 ? mean(cur.slice(Math.max(0, n - k - 1), n - 1).map((p) => p[1])) : cur[n - 1]![1];
  const ya = t.length > 1 ? mean(t.slice(1, k + 1).map((p) => p[1])) : t[0]![1];
  const ey = cur[n - 1]![1];
  return ey > yb + margin && ey > ya + margin;
}

/** Chain x-increasing segments where one's END ~ next's START — rebuilds a broken stroke, but refuses a valley
 *  join so two adjacent lobes meeting at the baseline are not merged into one full-width curve. */
export function chainCurves(segs: Pt[][], xtol = 4, ytol = 6): Pt[][] {
  const s = [...segs].sort((a, b) => a[0]![0] - b[0]![0]);
  const used = new Array(s.length).fill(false);
  const curves: Pt[][] = [];
  for (let i = 0; i < s.length; i++) {
    if (used[i]) continue;
    let cur = [...s[i]!];
    used[i] = true;
    let changed = true;
    while (changed) {
      changed = false;
      const [ex, ey] = cur[cur.length - 1]!;
      for (let j = 0; j < s.length; j++) {
        if (used[j]) continue;
        const t = s[j]!;
        if (Math.abs(t[0]![0] - ex) < xtol && Math.abs(t[0]![1] - ey) < ytol && !isValleyJoin(cur, t)) {
          cur = cur.concat(t);
          used[j] = true;
          changed = true;
          break;
        }
      }
    }
    curves.push(cur);
  }
  return curves;
}

/** Split a single-valued curve into separate lobes at DEEP inter-lobe valleys; keep shallow within-lobe dips
 *  (real double-humps) intact. h = -y (visual height). */
export function splitLobes(curve: Pt[], depthFrac = 0.35, minPts = 12): Pt[][] {
  const h = curve.map((p) => -p[1]);
  const hmin = Math.min(...h);
  for (let i = 0; i < h.length; i++) h[i] = h[i]! - hmin;
  const hmax = Math.max(...h);
  if (hmax <= 0) return [curve];
  const peaks: number[] = [];
  for (let i = 1; i < h.length - 1; i++) if (h[i]! >= h[i - 1]! && h[i]! >= h[i + 1]! && h[i]! > 0.15 * hmax) peaks.push(i);
  if (peaks.length < 2) return [curve];
  const splits: number[] = [];
  for (let p = 0; p < peaks.length - 1; p++) {
    const a = peaks[p]!, b = peaks[p + 1]!;
    let v = a;
    for (let i = a; i <= b; i++) if (h[i]! < h[v]!) v = i;
    if (h[v]! < depthFrac * Math.min(h[a]!, h[b]!)) splits.push(v);
  }
  if (!splits.length) return [curve];
  const out: Pt[][] = [];
  let prev = 0;
  for (const v of splits) {
    if (v - prev >= minPts) out.push(curve.slice(prev, v + 1));
    prev = v;
  }
  if (curve.length - prev >= minPts) out.push(curve.slice(prev));
  return out;
}
